src/main/java/com/conveyal/analysis/components/HttpApi.java

```diff
@@ -87,8 +88,18 @@ public HttpApi (
             // FIXME those internal endpoints should be hidden from the outside world by the reverse proxy.
             //       Or now with non-static Spark we can run two HTTP servers on different ports.
 
-            // Set CORS headers, to allow requests to this API server from any page.
-            res.header("Access-Control-Allow-Origin", "*");
+            // Set CORS headers, to allow requests to this API server from a frontend hosted on a different domain
+            // This used to be hardwired to Access-Control-Allow-Origin: * but that leaves the server open to XSRF
+            // attacks when authentication is disabled (e.g. when running locally).
+            if (config.accessControlAllowOrigin() == null || config.accessControlAllowOrigin().equals("null")) {
+                // Access-Control-Allow-Origin: null opens unintended security holes:
+                // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin
+                throw new IllegalArgumentException("Access-Control-Allow-Origin should not be null");
+            }
+
+            res.header("Access-Control-Allow-Origin", config.accessControlAllowOrigin());
+            // for caching, signal to the browser that responses may be different based on origin
```

Comment:
I guess it is present because we can and sometimes will set the header to *.

Reply:
It's been a while since I looked at this. I'm not sure if it's a security issue, but the HTTP spec requires they vary header when the response may vary based on other headers, which is the case any time Access-Control-Allow-Origin != * (or missing, I guess). From a convenience perspective, I guess it could come up if CORS was configured incorrectly and the browser cached a CORS miss, and then you fixed the configuration but the miss was cached.